### PR TITLE
Prevent reloading page without user approval when loading a model from local.

### PR DESCRIPTION
**Pull request by rondoor124**
Hi Pablo.
I fixed https://github.com/bldrs-ai/Share/issues/149, please check this video, and give me feedback.

https://user-images.githubusercontent.com/92342281/211874058-44229875-c677-41e5-89e1-30e1a413e0f2.mp4

Thank you.

### CODE REVIEW

File: src/Containers/CadView.jsx [hunk 1]
```diff
@@ -468,7 +482,7 @@ export default function CadView({
         height: ' calc(100vh - calc(100vh - 100%))',
         minHeight: '-webkit-fill-available',
       },
-    }}data-model-ready={modelReady}
+    }} data-model-ready={modelReady}
```
Comment (pablo-mayrgundter) [inline src/Containers/CadView.jsx:485]: good catch.   Please move both sx and data-model-ready onto their own lines

File: src/Containers/CadView.jsx [hunk 2]
```diff
@@ -193,6 +193,17 @@ export default function CadView({
...
     setAlert(<Alert onCloseCb={() => navToDefault(navigate, appPrefix)} message={msg}/>)
 
+  /**
+   * Prevent reloading page without user approval when loading a model from local
+   *
+   * @param {Event} e
+   * @return {string} confirmationMessage
+   */
+  function handleBeforeUnload(e) {
+    const confirmationMessage = 'It looks like you have been editing something. If you leave before saving, your changes will be lost.';
```
Comment (pablo-mayrgundter) [inline src/Containers/CadView.jsx:203]: Please use: 'Uploaded models are not saved.  If you leave this page you will need to upload the model again.'

**Review by pablo-mayrgundter — CHANGES_REQUESTED** [review]

**Comment by pablo-mayrgundter** [comment]
Actually, I see an error..  if you reload and accept, you go back to the index model, but reloading after that also brings up the warning, which should be cleared

**Comment by pablo-mayrgundter** [comment]
Also please add a test if it's straight-forward.. i think this would be approachable with a mock

File: src/Containers/CadView.jsx (hunk 1 above)
Comment (rondoor124) [reply src/Containers/CadView.jsx]: I have done.

File: src/Containers/CadView.jsx (hunk 2 above)
Comment (rondoor124) [reply src/Containers/CadView.jsx]: I have done.

**Comment by rondoor124** [comment]
I finished to not show warning message after reloading page.

**Comment by OlegMoshkovich** [comment]
image

When I try to navigate away from the page this message pops up.
This is should not happen.
@rondoor124

**Comment by rondoor124** [comment]
Hi @OlegMoshkovich, thank you for your feedback.
It seems better to show that warning message when leaving the site to let the user think seriously about unsaved data in the future, would like to know your suggestion.

**Comment by OlegMoshkovich** [comment]
@rondoor124 let's show the message only if the local model is loaded.
Currently, the message is always shown when the user navigates away from the site.

**Comment by rondoor124** [comment]
https://user-images.githubusercontent.com/92342281/212199690-a69ba261-231a-4d7f-bf34-414f2faedee3.mp4

@OlegMoshkovich, I checked the project again, everything works well as you mentioned, please check again, maybe need to clear cache. :)

File: src/utils/event.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Prevent reloading page without user approval
+ *
+ * @param {Event} e
+ * @return {string}
+ */
+export function handleBeforeUnload(e) {
+  const confirmationMessage = 'Uploaded models are not saved. If you leave this page you will need to upload the model again.';
```
Comment (pablo-mayrgundter) [inline src/utils/event.js:8]: Hey Ron, I'm actually not seeing this message displayed in any browser I try on Mac (Chrome, FF, Safari).. have you seen it, and if so on what config?

File: src/Containers/CadView.jsx [hunk 4]
```diff
@@ -559,6 +568,11 @@ export default function CadView({
         {alert}
       </>
       <SideDrawerWrapper/>
+      <input
```
Comment (pablo-mayrgundter) [inline src/Containers/CadView.jsx:571]: Per convo, please add/remove dynamically for only local files

File: src/Containers/CadView.test.jsx [hunk 5]
```diff
@@ -1,5 +1,6 @@
 import React, {useState} from 'react'
 import {render, renderHook, act, fireEvent, screen, waitFor} from '@testing-library/react'
+// import user from '@testing-library/user-event'
```
Comment (pablo-mayrgundter) [inline src/Containers/CadView.test.jsx:3]: remove or uncomment if needed

**Review by pablo-mayrgundter — CHANGES_REQUESTED** [review]

File: src/utils/event.js (hunk 3 above)
Comment (rondoor124) [reply src/utils/event.js]: as a matter of fact, the user approval prompt feature when exiting or reloading page can be implemented using `beforeunload` event of windows, no another way, so i think we can't make our own prompt dialog, hope you understand.

File: src/utils/file.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+/**
+ * @param {string} filepath
+ * @return {string}
+ */
+export function getNewModelRealPath(filepath) {
```
[the remark this answers is not in the document]
Comment (rondoor124) [reply src/utils/file.js]: i have done.

File: src/Containers/CadView.test.jsx [hunk 7]
```diff
@@ -103,4 +104,46 @@ describe('CadView', () => {
...
+            installPrefix=''
+            appPrefix=''
+            pathPrefix=''
+            modelPath={modelPath}
+            getNewModelRealPath={getModelPath2}
+          />
+        </ShareMock>,
+    )
+    expect(getModelPath2).not.toHaveBeenCalled()
+    expect(window.addEventListener).toHaveBeenCalledWith('beforeunload', expect.anything())
```
[the remark this answers is not in the document]
Comment (rondoor124) [reply src/Containers/CadView.test.jsx:143]: yes, you're right.

File: src/Containers/CadView.jsx (hunk 4 above)
Comment (rondoor124) [reply src/Containers/CadView.jsx]: i have done.

File: src/Containers/CadView.jsx [hunk 8]
```diff
@@ -589,7 +603,7 @@ function initViewer(pathPrefix, backgroundColorStr = '#abcdef') {
 
   // Highlight items when hovering over them
   window.onmousemove = (event) => {
-    v.prePickIfcItem()
+    // v.prePickIfcItem()
```
[the remark this answers is not in the document]
Comment (rondoor124) [reply src/Containers/CadView.jsx:604]: i have done.

File: src/Containers/CadView.test.jsx (hunk 5 above)
Comment (rondoor124) [reply src/Containers/CadView.test.jsx:3]: i have done.

File: src/utils/event.js (hunk 3 above)
Comment (pablo-mayrgundter) [reply src/utils/event.js]: Yeah, that part I see what you're saying.  I was thinking maybe a dialog before that, but I'm happy with PR as-is for now.

**Pull request merged**